app/uk.gov.hmrc.helptosavestridefrontend/controllers/StrideController.scala

```diff
@@ -37,63 +37,86 @@ import scala.concurrent.Future
+    Ok(views.html.get_eligibility_page(GiveNINOForm.giveNinoForm)).withSession(newSession)
   }(routes.StrideController.getEligibilityPage())
 
   def checkEligibilityAndGetPersonalInfo: Action[AnyContent] = authorisedFromStride { implicit request ⇒
-    GiveNINOForm.giveNinoForm.bindFromRequest().fold(
-      withErrors ⇒ Ok(views.html.get_eligibility_page(withErrors)),
-      form ⇒ {
-        val ninoEncoded = new String(base64Encode(form.nino))
-
-        val r = for {
-          eligibility ← helpToSaveConnector.getEligibility(ninoEncoded)
-          personalDetails ← getPersonalDetails(eligibility, ninoEncoded)
-        } yield eligibility -> personalDetails
-
-        r.fold(
-          error ⇒ {
-            logger.warn(s"error during get eligibility result and paye-personal-info, error: $error")
-            internalServerError()
-          },
-          {
-            case (Eligible(_), Some(details)) ⇒
-              Ok(views.html.you_are_eligible(details))
-            case (Ineligible(_), _) ⇒
-              SeeOther(routes.StrideController.youAreNotEligible().url)
-            case (AlreadyHasAccount(_), _) ⇒
-              SeeOther(routes.StrideController.accountAlreadyExists().url)
-            case _ ⇒
-              logger.warn("unknown error during checking eligibility and pay-personal-details")
-              internalServerError()
-          }
-        )
 
-      })
+    checkSession(
+      GiveNINOForm.giveNinoForm.bindFromRequest().fold(
+        withErrors ⇒ Ok(views.html.get_eligibility_page(withErrors)),
+        form ⇒ {
+          val ninoEncoded = new String(base64Encode(form.nino))
+          val r = for {
+            eligibility ← helpToSaveConnector.getEligibility(ninoEncoded)
+            personalDetails ← getPersonalDetails(eligibility, ninoEncoded)
+            keyWithSession ← storeDetails(eligibility, personalDetails)
+          } yield (eligibility, personalDetails, keyWithSession)
+
+          r.fold(
+            error ⇒ {
+              logger.warn(s"error during get eligibility result and paye-personal-info, error: $error")
+              internalServerError()
+            },
+            {
+              case (Eligible(_), Some(details), Some(ks)) ⇒
+                Ok(views.html.you_are_eligible(details)).withSession(ks.session)
+              case (Ineligible(_), _, _) ⇒
+                SeeOther(routes.StrideController.youAreNotEligible().url)
+              case (AlreadyHasAccount(_), _, _) ⇒
+                SeeOther(routes.StrideController.accountAlreadyExists().url)
+              case _ ⇒
+                logger.warn("unknown error during checking eligibility and pay-personal-details")
+                internalServerError()
+            }
+          )
+        })
+    )(handle)
   }(routes.StrideController.checkEligibilityAndGetPersonalInfo())
 
   def youAreNotEligible: Action[AnyContent] = authorisedFromStride { implicit request ⇒
     Ok(views.html.you_are_not_eligible())
```

Comment:
we may want to do `checkSession` here as well to make sure they're really ineligible

Reply:
lets discuss